File: csharp/annalyns-infiltration/AnnalynsInfiltration.cs
using System;

static class QuestLogic
{
    public static bool CanFastAttack(bool knightIsAwake) =>
        !knightIsAwake;

    public static bool CanSpy(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake) =>
        prisonerIsAwake || knightIsAwake || archerIsAwake;

    public static bool CanSignalPrisoner(bool archerIsAwake, bool prisonerIsAwake) =>
        prisonerIsAwake && !archerIsAwake;

    public static bool CanFreePrisoner(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake, bool petDogIsPresent) =>
        prisonerIsAwake && !knightIsAwake && !archerIsAwake ||
        petDogIsPresent && !archerIsAwake;
}

File: r/annalyns-infiltration/annalyns-infiltration.R
can_fast_attack <- function(knight_is_awake) {
  !knight_is_awake
}

can_spy <- function(knight_is_awake, archer_is_awake, prisoner_is_awake) {
  knight_is_awake || archer_is_awake || prisoner_is_awake
}

can_signal_prisoner <- function(archer_is_awake, prisoner_is_awake) {
  prisoner_is_awake && !archer_is_awake
}

can_free_prisoner <- function(knight_is_awake,
                              archer_is_awake,
                              prisoner_is_awake,
                              pet_dog_is_present) {
  pet_dog_is_present && !archer_is_awake ||
    prisoner_is_awake && !archer_is_awake && !knight_is_awake
}
